### PR TITLE
Add data structures for bucket table writer support

### PR DESCRIPTION
**Pull request by xiaoxmeng**
Add HiveBucketProperty and HiveSortingColumn and data structures
to support table bucket write. HiveBucketProperty specifies the bucket
function and bucket columns. HiveSortingColumn specifies the sorting
column and orders if required.

### CODE REVIEW

File: velox/core/PlanNode.h [hunk 1]
```diff
@@ -69,6 +69,11 @@ class SortOrder {
     return nullsFirst_;
   }
 
+  bool operator==(const SortOrder& other) const {
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:72]: nit: Would be nice to add a small test.

File: velox/connectors/hive/HiveDataSink.h [hunk 2]
```diff
@@ -79,6 +79,104 @@ class LocationHandle : public ISerializable {
...
+  static void registerSerDe();
+
+ private:
+  const std::string sortColumn_;
+  const core::SortOrder sortOrder_;
+};
+
+class HiveBucketProperty : public ISerializable {
+ public:
+  enum class Kind { HIVE_COMPATIBLE, PRESTO_NATIVE };
```
Comment (mbasmanova) [inline velox/connectors/hive/HiveDataSink.h:113]: naming: kHiveCompatible and kPrestoNative

File: velox/connectors/hive/HiveDataSink.h [hunk 3]
```diff
@@ -79,6 +79,104 @@ class LocationHandle : public ISerializable {
...
+
+  HiveBucketProperty(
+      Kind kind,
+      int32_t bucketCount,
+      const std::vector<std::string>& bucketedBy,
+      const std::vector<TypePtr>& bucketedTypes,
+      const std::vector<std::shared_ptr<const HiveSortingColumn>>& sortedBy);
```
Comment (mbasmanova) [inline velox/connectors/hive/HiveDataSink.h:120]: Why not just std::vector<HiveSortingColumn> ? Do we need to wrap HiveSortingColumn in a shared pointer?

**Review by mbasmanova — APPROVED** [review]
@xiaoxmeng Thanks.

File: velox/connectors/hive/HiveDataSink.h (hunk 3 above)
Comment (xiaoxmeng) [reply velox/connectors/hive/HiveDataSink.h:120]: It is ease dynamic object deserialization handling. Currently we only support shared_ptr/unique_ptr object creation through deserialization.

Comment (mbasmanova) [reply velox/connectors/hive/HiveDataSink.h:120]: Is this really so? I don't think such limitation exists.

Comment (mbasmanova) [reply velox/connectors/hive/HiveDataSink.h:120]: See WindowNode for an example.

**Pull request closed without merging**